File: test/hardware/unit/test_l1_arbiter.py
import cocotb
from cocotb.triggers import RisingEdge, Timer
from cocotb.clock import Clock
from pathlib import Path
from test.driver import run_hardware_test
from test.env import get_rtl_dir


@cocotb.test()
async def test_l1_arbiter_basic(dut):
    """Test L1 arbiter with simple read requests."""
    
    clock = Clock(dut.clk, 10, units="ns")
    cocotb.start_soon(clock.start())
    
    # Reset
    dut.rst_n.value = 0
    dut.icache_req.value = 0
    dut.dcache_req.value = 0
    dut.dcache_we.value = 0
    dut.m_ready.value = 0
    await Timer(20, units="ns")
    await RisingEdge(dut.clk)
    dut.rst_n.value = 1
    await RisingEdge(dut.clk)
    
    # Test 1: I-Cache request
    dut.icache_addr.value = 0x1000
    dut.icache_req.value = 1
    dut.m_rdata.value = 0xDEADBEEF
    await RisingEdge(dut.clk)
    
    # Arbiter should forward request
    assert dut.m_req.value == 1, "m_req should be high"
    assert dut.m_addr.value == 0x1000, "Address should be forwarded"
    
    # Simulate memory ready
    dut.m_ready.value = 1
    await RisingEdge(dut.clk)
    
    # Check response
    assert dut.icache_ready.value == 1, "icache_ready should be high"
    assert dut.icache_rdata.value == 0xDEADBEEF, "Data should be forwarded"
    
    # Release
    dut.icache_req.value = 0
    dut.m_ready.value = 0
    await RisingEdge(dut.clk)
    await RisingEdge(dut.clk)
    
    # Test 2: D-Cache request (priority)
    dut.dcache_addr.value = 0x2000
    dut.dcache_req.value = 1
    dut.dcache_we.value = 0
    dut.icache_addr.value = 0x3000
    dut.icache_req.value = 1
    await RisingEdge(dut.clk)
    
    # D-Cache should have priority
    assert dut.m_addr.value == 0x2000, "D-Cache should have priority"
    
    dut.m_ready.value = 1
    await RisingEdge(dut.clk)
    
    dut.dcache_req.value = 0
    dut.m_ready.value = 0
    await RisingEdge(dut.clk)
    
    # Now I-Cache should get access
    assert dut.m_addr.value == 0x3000, "I-Cache should get access now"
    
    dut._log.info("L1 Arbiter test passed")


def test_l1_arbiter():
    rtl_dir = get_rtl_dir()
    run_hardware_test(
        module_name=Path(__file__).stem,
        toplevel="l1_arbiter",
        verilog_sources=[rtl_dir / "cache" / "l1_arbiter.v"]
    )

File: test/unit_test/test_l1_arbiter.cpp
#define DOCTEST_CONFIG_IMPLEMENT_WITH_MAIN
#include "doctest.h"
#include "tb_base.h"
#include "Vl1_arbiter.h"
#include <string>

class L1ArbiterTestbench : public ClockedTestbench<Vl1_arbiter> {
public:
    L1ArbiterTestbench() : ClockedTestbench<Vl1_arbiter>(100, false) {
        // Initialize inputs
        dut->icache_req = 0;
        dut->icache_addr = 0;
        dut->dcache_req = 0;
        dut->dcache_addr = 0;
        dut->dcache_wdata = 0;
        dut->dcache_we = 0;
        dut->dcache_be = 0;
        dut->m_ready = 0;
        dut->m_rdata = 0;
    }
    
    void set_clk(uint8_t value) override {
        dut->clk = value;
    }
    
    void reset() {
        dut->rst_n = 0;
        tick();
        dut->rst_n = 1;
        tick();
    }
    
    void test_icache_request() {
        
        // I-Cache requests data
        dut->icache_addr = 0x1000;
        dut->icache_req = 1;
        dut->m_rdata = 0xDEADBEEF;
        tick();  // Enter STATE_ICACHE
        
        // Arbiter should forward request
        CHECK(dut->m_req == 1);
        CHECK(dut->m_addr == 0x1000);
        
        // Simulate memory ready (stay in STATE_ICACHE until ready)
        dut->m_ready = 1;
        eval();  // Combinational response
        
        // Check response
        CHECK(dut->icache_ready == 1);
        CHECK(dut->icache_rdata == 0xDEADBEEF);
        
        // Next cycle goes back to IDLE
        tick();
        
        // Release
        dut->icache_req = 0;
        dut->m_ready = 0;
        tick();
    }
    
    void test_dcache_priority() {
        
        // Both request simultaneously
        dut->dcache_addr = 0x2000;
        dut->dcache_req = 1;
        dut->dcache_we = 0;
        dut->icache_addr = 0x3000;
        dut->icache_req = 1;
        tick();  // Enter STATE_DCACHE (priority)
        
        // D-Cache should have priority
        CHECK(dut->m_addr == 0x2000);
        
        // Complete D-Cache request
        dut->m_ready = 1;
        dut->m_rdata = 0x11111111;
        eval();
        CHECK(dut->dcache_ready == 1);
        tick();  // Back to IDLE
        
        dut->dcache_req = 0;
        dut->m_ready = 0;
        tick();  // Enter STATE_ICACHE now
        
        // Now I-Cache should get access
        CHECK(dut->m_addr == 0x3000);
        
        // Cleanup
        dut->m_ready = 1;
        eval();
        tick();
        dut->icache_req = 0;
        dut->m_ready = 0;
        tick();
    }
    
    void test_dcache_write() {
        
        // D-Cache write
        dut->dcache_addr = 0x4000;
        dut->dcache_wdata = 0x12345678;
        dut->dcache_we = 1;
        dut->dcache_be = 0b1111;
        dut->dcache_req = 1;
        tick();  // Enter STATE_DCACHE
        
        // Check forwarding
        CHECK(dut->m_req == 1);
        CHECK(dut->m_addr == 0x4000);
        CHECK(dut->m_wdata == 0x12345678);
        CHECK(dut->m_we == 1);
        CHECK(dut->m_be == 0b1111);
        
        // Complete write
        dut->m_ready = 1;
        eval();
        CHECK(dut->dcache_ready == 1);
        
        // Back to IDLE
        tick();
        
        // Cleanup
        dut->dcache_req = 0;
        dut->dcache_we = 0;
        dut->m_ready = 0;
        tick();
    }
};

TEST_CASE("L1 Arbiter") {
L1ArbiterTestbench tb;
        
        tb.reset();
        tb.test_icache_request();
        tb.test_dcache_priority();
        tb.test_dcache_write();
}
